s3/pkg/service/multipart.go

```diff
@@ -446,6 +504,66 @@ func (s *s3Service) AbortMultipartUpload(ctx context.Context, in *pb.AbortMultip
 
 func (s *s3Service) ListObjectParts(ctx context.Context, in *pb.ListObjectPartsRequest, out *pb.ListObjectPartsResponse) error {
 	log.Info("ListObjectParts is called in s3 service.")
+	var err error
+	defer func() {
+		out.ErrorCode = GetErrCode(err)
+	}()
+
+	bucketName := in.BucketName
+	objectKey := in.ObjectKey
+	uploadId := in.UploadId
+
+	multipart, err := s.MetaStorage.GetMultipart(bucketName, objectKey, uploadId)
+	if err != nil {
+		log.Errorln("failed to get multipart info. err:", err)
+		return err
+	}
+
+	isAdmin, tenantId, _, err := util.GetCredentialFromCtx(ctx)
+	if err != nil && isAdmin == false {
+		log.Error("get tenant id failed")
+		err = ErrInternalError
+		return err
+	}
+
+	if !isAdmin {
+		switch multipart.Metadata.Acl.CannedAcl {
+		case "public-read", "public-read-write":
+			break
+		default:
+			if multipart.Metadata.TenantId != tenantId {
+				err = ErrAccessDenied
+				return err
+			}
+		}
+	}
+
+	out.Initiator = &pb.Owner{Id: multipart.Metadata.InitiatorId, DisplayName: multipart.Metadata.InitiatorId}
+	out.Owner = &pb.Owner{Id: multipart.Metadata.TenantId, DisplayName: multipart.Metadata.TenantId}
+	out.MaxParts = int64(in.MaxParts)
+	out.Parts = make([]*pb.Part, 0)
+	for i := in.PartNumberMarker + 1; i <= MAX_PART_NUMBER; i++ {
+		if p, ok := multipart.Parts[int(i)]; ok {
+			out.Parts = append(out.Parts, &pb.Part{
+				PartNumber:   i,
+				ETag:         "\"" + p.Etag + "\"",
+				Size:         p.Size,
+				LastModified: p.LastModified,
+			})
+
+			if int64(len(out.Parts)) > in.MaxParts {
+				break
+			}
+		}
+	}
+	if int64(len(out.Parts)) == in.MaxParts+1 {
+		out.IsTruncated = true
+		out.NextPartNumberMarker = out.Parts[out.MaxParts].PartNumber
```

Comment:
Does NextPartNumberMarker equals the PartNumber of current MaxPart?

Reply:
thanks for your reviewing, yes, it is. if there are more than maxPart items, it indicate that IsTruncated  should be true,  and set NextPartNumberMarker as  the PartNumber of current MaxPart.